File: packages/webfetch/src/fence.ts
import { toolError, type ToolError } from "@agent-sh/harness-core";
import type {
  WebFetchMethod,
  WebFetchSessionConfig,
} from "./types.js";

/**
 * Permission hook call for webfetch. Mirrors the shape used by read /
 * grep / glob / bash but with WebFetch-specific metadata. Returns a
 * decision string; "ask" is treated as "deny" in autonomous mode.
 */
export async function askPermission(
  session: WebFetchSessionConfig,
  args: {
    method: WebFetchMethod;
    url: string;
    host: string;
    bodyBytes: number;
    headerKeys: readonly string[];
    extract: string;
    timeoutMs: number;
    maxRedirects: number;
  },
): Promise<
  | { decision: "allow" | "allow_once" }
  | { decision: "deny"; reason: string }
> {
  const { permissions } = session;
  const pattern = `WebFetch(domain:${args.host})`;

  if (permissions.hook === undefined) {
    if (permissions.unsafeAllowFetchWithoutHook === true) {
      return { decision: "allow" };
    }
    return {
      decision: "deny",
      reason:
        "webfetch tool has no permission hook configured; refusing to fetch untrusted URLs. Wire a hook or set session.permissions.unsafeAllowFetchWithoutHook for test fixtures.",
    };
  }

  const decision = await permissions.hook({
    tool: "webfetch",
    path: args.url,
    action: "read",
    always_patterns: [pattern],
    metadata: {
      method: args.method,
      url: args.url,
      host: args.host,
      body_bytes: args.bodyBytes,
      headers_sent: args.headerKeys,
      extract: args.extract,
      timeout_ms: args.timeoutMs,
      redirect_limit: args.maxRedirects,
    },
  });
  if (decision === "deny") {
    return {
      decision: "deny",
      reason: `URL blocked by permission policy. Pattern hint: ${pattern}`,
    };
  }
  if (decision === "allow" || decision === "allow_once") {
    return { decision };
  }
  return {
    decision: "deny",
    reason:
      "Permission hook returned 'ask' but webfetch runs in autonomous mode. Configure the hook to return allow or deny.",
  };
}

export function permissionDeniedError(
  url: string,
  reason: string,
): ToolError {
  const echoUrl = url.length > 300 ? url.slice(0, 300) + "..." : url;
  return toolError(
    "PERMISSION_DENIED",
    `${reason}\nURL: ${echoUrl}`,
    { meta: { url } },
  );
}

File: crates/webfetch/src/fence.rs
use harness_core::permissions::PermissionQuery;
use harness_core::{PermissionDecision, ToolError, ToolErrorCode};

use crate::types::{WebFetchMethod, WebFetchSessionConfig};

#[derive(Debug, Clone)]
pub enum PermissionOutcome {
    Allow,
    Deny { reason: String },
}

pub struct AskArgs<'a> {
    pub method: WebFetchMethod,
    pub url: &'a str,
    pub host: &'a str,
    pub body_bytes: usize,
    pub header_keys: Vec<String>,
    pub extract: &'a str,
    pub timeout_ms: u64,
    pub max_redirects: u32,
}

pub async fn ask_permission(
    session: &WebFetchSessionConfig,
    args: AskArgs<'_>,
) -> PermissionOutcome {
    let permissions = &session.permissions.inner;
    let pattern = format!("WebFetch(domain:{})", args.host);

    if permissions.hook.is_none() {
        if session.permissions.unsafe_allow_fetch_without_hook {
            return PermissionOutcome::Allow;
        }
        return PermissionOutcome::Deny {
            reason: "webfetch tool has no permission hook configured; refusing to fetch untrusted URLs. Wire a hook or set permissions.unsafe_allow_fetch_without_hook for test fixtures."
                .to_string(),
        };
    }

    let metadata = serde_json::json!({
        "method": args.method.as_str(),
        "url": args.url,
        "host": args.host,
        "body_bytes": args.body_bytes,
        "headers_sent": args.header_keys,
        "extract": args.extract,
        "timeout_ms": args.timeout_ms,
        "redirect_limit": args.max_redirects,
    });
    let query = PermissionQuery {
        tool: "webfetch".to_string(),
        path: args.url.to_string(),
        action: "read".to_string(),
        always_patterns: vec![pattern.clone()],
        metadata,
    };
    let hook = permissions.hook.as_ref().unwrap();
    let decision = (hook)(query).await;
    match decision {
        PermissionDecision::Allow | PermissionDecision::AllowOnce => PermissionOutcome::Allow,
        PermissionDecision::Deny => PermissionOutcome::Deny {
            reason: format!("URL blocked by permission policy. Pattern hint: {}", pattern),
        },
        PermissionDecision::Ask => PermissionOutcome::Deny {
            reason: "Permission hook returned 'ask' but webfetch runs in autonomous mode. Configure the hook to return allow or deny.".to_string(),
        },
    }
}

pub fn permission_denied_error(url: &str, reason: &str) -> ToolError {
    let echo_url = if url.len() > 300 {
        format!("{}...", &url[..300])
    } else {
        url.to_string()
    };
    ToolError::new(
        ToolErrorCode::PermissionDenied,
        format!("{}\nURL: {}", reason, echo_url),
    )
    .with_meta(serde_json::json!({ "url": url }))
}
